File: engine/max_engine/postprocess.py
"""Output post-processing for DSL command responses.

For streaming responses the extension applies these transforms incrementally on
the accumulated text (see extension/src/extension.ts postProcess).  This module
provides the same logic for non-streaming callers (tests, CLI, future REST
clients that request a single buffered response).
"""

from __future__ import annotations

import re

_OPEN_FENCE = re.compile(r"^```[^\n]*\n?", re.MULTILINE)
_CLOSE_FENCE = re.compile(r"\n?```\s*$")


def strip_fences(text: str) -> str:
    """Remove the outermost opening and closing markdown code fences."""
    text = _OPEN_FENCE.sub("", text, count=1)
    text = _CLOSE_FENCE.sub("", text)
    return text


def reindent(text: str, base_indent: str) -> str:
    """Prepend *base_indent* to every line except the first.

    The first line is placed at the cursor which is already indented; subsequent
    lines need the base indent added so the whole block aligns correctly.
    Blank continuation lines are left blank (no trailing whitespace injected).
    """
    if not base_indent:
        return text
    lines = text.split("\n")
    result = [
        line if (i == 0 or line == "") else base_indent + line
        for i, line in enumerate(lines)
    ]
    return "\n".join(result)


def postprocess(text: str, base_indent: str = "") -> str:
    """Full pipeline: strip fences then reindent."""
    return reindent(strip_fences(text), base_indent)

File: clients/nvim/lua/max/postprocess.lua
-- Strip opening/closing markdown code fences and re-apply the command line's
-- indentation to continuation lines. Applied to the *accumulated* stream buffer
-- on every chunk so a fence never appears in the buffer even mid-stream.
--
-- base_indent: whitespace prefix of the command line — prepended to every line
-- after the first so the whole block sits at the right column.

local M = {}

-- Split on "\n", keeping empty trailing segments (so a trailing newline shows).
local function split_lines(text)
  local lines = {}
  local start = 1
  while true do
    local nl = text:find("\n", start, true)
    if not nl then
      lines[#lines + 1] = text:sub(start)
      break
    end
    lines[#lines + 1] = text:sub(start, nl - 1)
    start = nl + 1
  end
  return lines
end

function M.process(text, base_indent)
  base_indent = base_indent or ""
  local lines = split_lines(text)

  -- Strip opening fence (first line begins with ```).
  local first = 1
  if lines[1] ~= nil and lines[1]:match("^```") then
    first = 2
  end

  -- Strip closing fence (last non-empty line is ``` only).
  local last = #lines
  while last > first and (lines[last] or ""):match("^%s*$") do
    last = last - 1
  end
  if last >= first and (lines[last] or ""):match("^```%s*$") then
    last = last - 1
  end

  -- Re-indent: first line sits at the cursor (already indented); later lines get
  -- base_indent prepended. Blank lines stay blank.
  local out = {}
  for i = first, last do
    local line = lines[i]
    if i == first or line == "" or base_indent == "" then
      out[#out + 1] = line
    else
      out[#out + 1] = base_indent .. line
    end
  end
  return table.concat(out, "\n")
end

return M
